File: cairn/adapters/asana.py
#!/usr/bin/env python3
"""Asana adapter for cairn (REST 1.0, stdlib only).

config (sync.json):
{
  "project_gid": "1209…",        # required — tasks are created in this project
  "token_env": "ASANA_TOKEN"     # env var holding a Personal Access Token
}

Auth: Bearer Personal Access Token. Create one at
https://app.asana.com/0/my-apps and export it. Asana tasks have no native
"in progress", so push maps in_progress->open (incomplete) and pull maps
completed->closed / incomplete->open.
"""
import json
import os
import sys
import urllib.error
import urllib.request

BASE = "https://app.asana.com/api/1.0"


def token(cfg):
    t = os.environ.get(cfg.get("token_env", "ASANA_TOKEN"), "")
    if not t:
        print(f"asana adapter: missing token env var "
              f"({cfg.get('token_env','ASANA_TOKEN')})", file=sys.stderr)
        sys.exit(1)
    return t


def api(cfg, method, path, body=None):
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(BASE + path, data=data, method=method)
    req.add_header("Authorization", "Bearer " + token(cfg))
    req.add_header("Content-Type", "application/json")
    try:
        with urllib.request.urlopen(req) as r:
            raw = r.read().decode()
            return json.loads(raw).get("data", {}) if raw else {}
    except urllib.error.HTTPError as e:
        print(f"asana {method} {path} -> {e.code}: {e.read().decode()[:300]}",
              file=sys.stderr)
        sys.exit(1)


def push(event, cfg):
    ext = event.get("external_id")
    action = event["action"]
    if action == "create" or (action == "update" and not ext):
        d = api(cfg, "POST", "/tasks", {"data": {
            "name": event["title"], "notes": event["body"] or "",
            "projects": [cfg["project_gid"]],
            "completed": event["status"] == "closed",
        }})
        return str(d.get("gid", ""))
    if action == "update":
        api(cfg, "PUT", f"/tasks/{ext}", {"data": {
            "name": event["title"], "notes": event["body"] or "",
            "completed": event["status"] == "closed",
        }})
        return ext
    if action == "close":
        if ext:
            api(cfg, "PUT", f"/tasks/{ext}", {"data": {"completed": True}})
        return ext or ""
    print(f"unknown action {action}", file=sys.stderr)
    sys.exit(1)


def pull(cfg, items):
    out = []
    for it in items:
        ext = it.get("external_id")
        if not ext:
            continue
        try:
            d = api(cfg, "GET",
                    f"/tasks/{ext}?opt_fields=name,notes,completed,modified_at")
        except SystemExit:
            continue
        out.append({
            "bd_id": it["bd_id"], "external_id": str(d.get("gid", ext)),
            "title": d.get("name", ""), "body": d.get("notes", "") or "",
            "status": "closed" if d.get("completed") else "open",
            "updated_at": d.get("modified_at"),
        })
    return out


def main():
    event = json.load(sys.stdin)
    cfg = event.get("config", {})
    if event["action"] != "pull" and "project_gid" not in cfg:
        print("asana adapter: config.project_gid is required", file=sys.stderr)
        sys.exit(1)
    if event["action"] == "pull":
        print(json.dumps(pull(cfg, event.get("items", []))))
    else:
        print(push(event, cfg))


if __name__ == "__main__":
    main()

File: cairn/server/src/tracker/adapters/asana.ts
import { z } from "zod";
import { CairnError } from "../../errors.js";
import { fetchJson, type FetchLike } from "../http.js";
import type {
  Capability, Issue, IssueCreate, IssuePatch, IssueState, Phase, Tracker,
} from "../types.js";

const API = "https://app.asana.com/api/1.0";
const OPT_FIELDS = "name,notes,completed,modified_at,memberships.section.gid";
const LIST_CAP = 100;

export const configSchema = z.object({
  projectGid: z.string().regex(/^\d+$/),
  tokenEnv: z.string().default("ASANA_TOKEN"),
});

export function make(config: z.infer<typeof configSchema>, fetchImpl?: FetchLike): Tracker {
  return new AsanaTracker(config, fetchImpl);
}

interface AsanaTask {
  gid: string; name: string; notes: string | null; completed: boolean;
  modified_at: string;
  memberships?: Array<{ section?: { gid: string } }>;
}

interface AsanaSection {
  gid: string; name: string;
}

export class AsanaTracker implements Tracker {
  readonly capabilities: Capability = {
    hasInProgress: false, hasPhases: true, hasDependencies: true, hasLabels: false,
  };

  constructor(
    private readonly cfg: { projectGid: string; tokenEnv: string },
    private readonly fetchImpl: FetchLike = fetch,
    private readonly tokenProvider: () => string = () => resolveAsanaToken(cfg.tokenEnv),
  ) {}

  private headers(): Record<string, string> {
    return {
      authorization: `Bearer ${this.tokenProvider()}`,
      "content-type": "application/json",
    };
  }

  private async api(method: string, path: string, body?: unknown, context?: string): Promise<unknown> {
    const raw = await fetchJson(this.fetchImpl, `${API}${path}`, {
      method,
      headers: this.headers(),
      body: body === undefined ? undefined : JSON.stringify(body),
    }, { context: context ? `asana ${context}` : "asana" });
    return (raw as { data?: unknown }).data;
  }

  private assertId(id: string): void {
    if (!/^\d+$/.test(id)) {
      throw new CairnError("NOT_FOUND", `invalid task id: ${id}`,
        "task id must be a numeric gid");
    }
  }

  private normalize(raw: AsanaTask): Issue {
    const sectionGid = raw.memberships?.[0]?.section?.gid;
    return {
      id: raw.gid, title: raw.name, body: raw.notes ?? "",
      state: raw.completed ? "closed" : "open",
      labels: [],
      phase: sectionGid,
      updatedAt: raw.modified_at,
      url: `https://app.asana.com/0/${this.cfg.projectGid}/${raw.gid}`,
    };
  }

  async createIssue(input: IssueCreate): Promise<Issue> {
    if (input.phase && !/^\d+$/.test(input.phase)) {
      throw new CairnError("CONFIG_INVALID",
        `invalid phase: ${input.phase}`,
        "phase must be a numeric section gid");
    }
    const body = {
      data: {
        name: input.title, notes: input.body ?? "",
        projects: [this.cfg.projectGid],
      },
    };
    const raw = (await this.api("POST", "/tasks", body, "issue_create")) as AsanaTask;
    if (input.phase) {
      await this.api("POST", `/sections/${input.phase}/addTask`,
        { data: { task: raw.gid } }, "issue_assign_phase");
    }
    return this.normalize({ ...raw, memberships: input.phase
      ? [{ section: { gid: input.phase } }] : raw.memberships });
  }

  async getIssue(id: string): Promise<Issue> {
    this.assertId(id);
    const raw = await this.api("GET", `/tasks/${id}?opt_fields=${OPT_FIELDS}`,
      undefined, "issue_get");
    return this.normalize(raw as AsanaTask);
  }

  async updateIssue(id: string, patch: IssuePatch): Promise<Issue> {
    this.assertId(id);
    const body: Record<string, unknown> = {};
    if (patch.title !== undefined) body.name = patch.title;
    if (patch.body !== undefined) body.notes = patch.body;
    // Asana has no native in-progress state: in_progress and open both map
    // to completed:false. Never write completed:true except for an explicit close.
    if (patch.state === "closed") body.completed = true;
    if (patch.state === "open" || patch.state === "in_progress") body.completed = false;
    const raw = await this.api("PUT", `/tasks/${id}`, { data: body }, "issue_update");
    return this.normalize(raw as AsanaTask);
  }

  async closeIssue(id: string): Promise<Issue> {
    return this.updateIssue(id, { state: "closed" });
  }

  async listIssues(filter?: { phase?: string; state?: IssueState }): Promise<Issue[]> {
    let raw: AsanaTask[];
    if (filter?.phase) {
      if (!/^\d+$/.test(filter.phase)) {
        throw new CairnError("NOT_FOUND", `invalid phase: ${filter.phase}`,
          "phase must be a numeric section gid");
      }
      const list = await this.api("GET",
        `/sections/${filter.phase}/tasks?opt_fields=${OPT_FIELDS}`,
        undefined, "issue_list_phase");
      raw = list as AsanaTask[];
    } else {
      const list = await this.api("GET",
        `/projects/${this.cfg.projectGid}/tasks?opt_fields=${OPT_FIELDS}&limit=${LIST_CAP}`,
        undefined, "issue_list");
      raw = list as AsanaTask[];
      if (raw.length === LIST_CAP) {
        console.error(`[cairn] asana issue_list truncated at ${LIST_CAP} items for project ${this.cfg.projectGid}`);
      }
    }
    let issues = raw.map((r) => this.normalize(r));
    if (filter?.state) issues = issues.filter((i) => i.state === filter.state);
    return issues;
  }

  async createPhase(name: string): Promise<Phase> {
    const raw = (await this.api("POST", `/projects/${this.cfg.projectGid}/sections`,
      { data: { name } }, "phase_create")) as AsanaSection;
    return { id: raw.gid, name: raw.name, state: "open" };
  }

  async listPhases(): Promise<Phase[]> {
    const raw = (await this.api("GET", `/projects/${this.cfg.projectGid}/sections`,
      undefined, "phase_list")) as AsanaSection[];
    return raw.map((s) => ({ id: s.gid, name: s.name, state: "open" as const }));
  }
}

export function resolveAsanaToken(tokenEnv: string): string {
  const env = process.env[tokenEnv];
  if (env) return env;
  throw new CairnError("AUTH_MISSING", `no Asana credentials (${tokenEnv})`,
    `export ${tokenEnv}`);
}
